File: event-bus-rabbit/src/main/java/pegasus/eventbus/gson/GsonSerializer.java
package pegasus.eventbus.gson;

import java.io.UnsupportedEncodingException;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import pegasus.eventbus.amqp.Serializer;

public class GsonSerializer implements Serializer {

	Gson gson;
	
	public GsonSerializer() {
		gson = new GsonBuilder()
			.enableComplexMapKeySerialization()
			.create();
	}
	
	@Override
	public byte[] serialize(Object object) {
		String json = gson.toJson(object);
	
		try {
			return json.getBytes("UTF-8");
		} catch (UnsupportedEncodingException e) {
			throw new RuntimeException("Failed to encode json as UTF-8", e);
		}
	}

	@Override
	public <T> T deserialize(byte[] bytes, Class<? extends T> type) {
		String json;
		try {
			json = new String(bytes, "UTF-8");
		} catch (UnsupportedEncodingException e) {
			throw new RuntimeException("Failed to dencode json as UTF-8", e);
		}
		return gson.fromJson(json, type);
	}

}

File: event-bus-rabbit/src/main/dotnet/gson/GsonSerializer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

using log4net;
using Newtonsoft.Json;

using pegasus.eventbus.amqp;


namespace pegasus.eventbus.gson
{
    public class GsonSerializer : IEventSerializer
    {
        private static readonly ILog LOG = LogManager.GetLogger(typeof(GsonSerializer));


        public byte[] Serialize(object ev)
        {
            if (null == ev) throw new ArgumentNullException("ev");

            byte[] buffer;

            try
            {
                buffer = new UTF8Encoding().GetBytes(JsonConvert.SerializeObject(ev));
            }
            catch(Exception ex)
            {
                LOG.Error("Failed to serialize an event of type " + ev.GetType().FullName, ex);
                throw;
            }

            return buffer;
        }

        public TEvent Deserialize<TEvent>(byte[] rawEvent)
        {
            TEvent ev = default(TEvent);

            try
            {
                ev = JsonConvert.DeserializeObject<TEvent>(new UTF8Encoding().GetString(rawEvent));
            }
            catch (Exception ex)
            {
                LOG.Error("Failed to deserialize an event of type " + typeof(TEvent).FullName, ex);
                throw;
            }

            return ev;
        }
    }
}
